mypyc/emitmodule.py

```diff
@@ -127,23 +203,172 @@ def compile_modules_to_c(
             insert_ref_count_opcodes(fn)
 
-    source_paths = {module_name: result.files[module_name].path
-                    for module_name in module_names}
+    return modules
+
+
+def compile_modules_to_ir(
+    result: BuildResult,
+    mapper: genops.Mapper,
+    compiler_options: CompilerOptions,
+    errors: Errors,
+) -> ModuleIRs:
+    """Compile a collection of modules into ModuleIRs.
+
+    The modules to compile are specified as part of mapper's group_map.
+
+    Returns the IR of the modules.
+    """
+    deser_ctx = DeserMaps({}, {})
+    modules = {}
+
+    # Process the graph by SCC in topological order, like we do in mypy.build
+    for scc in sorted_components(result.graph):
+        scc_states = [result.graph[id] for id in scc]
+        trees = [st.tree for st in scc_states if st.id in mapper.group_map and st.tree]
+
+        if not trees:
+            continue
+
+        fresh = all(id not in result.manager.rechecked_modules for id in scc)
+        if fresh:
+            load_scc_from_cache(trees, result, mapper, deser_ctx)
+        else:
+            scc_ir = compile_scc_to_ir(trees, result, mapper, compiler_options, errors)
+            modules.update(scc_ir)
+
+    return modules
+
+
+def compile_ir_to_c(
+    groups: Groups,
+    modules: ModuleIRs,
+    result: BuildResult,
+    mapper: genops.Mapper,
+    compiler_options: CompilerOptions,
+) -> Dict[Optional[str], List[Tuple[str, str]]]:
+    """Compile a collection of ModuleIRs to C source text.
+
+    Returns a dictionary mapping group names to a list of (file name,
+    file text) pairs.
+    """
+    source_paths = {source.module: result.graph[source.module].xpath
+                    for sources, _ in groups for source in sources}
 
     names = NameGenerator([[source.module for source in sources] for sources, _ in groups])
 
     # Generate C code for each compilation group. Each group will be
     # compiled into a separate extension module.
-    ctext = []
+    ctext = {}  # type: Dict[Optional[str], List[Tuple[str, str]]]
     for group_sources, group_name in groups:
-        group_modules = [(source.module, modules[source.module]) for source in group_sources]
+        group_modules = [(source.module, modules[source.module]) for source in group_sources
+                         if source.module in modules]
+        if not group_modules:
+            ctext[group_name] = []
+            continue
         literals = mapper.literals[group_name]
         generator = GroupGenerator(
-            literals, group_modules, source_paths, group_name, group_map, names,
+            literals, group_modules, source_paths, group_name, mapper.group_map, names,
             compiler_options.multi_file
         )
-        ctext.append(generator.generate_c_for_modules())
-    return modules, ctext
+        ctext[group_name] = generator.generate_c_for_modules()
+
+    return ctext
+
+
+def get_ir_cache_name(id: str, path: str, options: Options) -> str:
```

Comment:
Are the contents of the new cache files documented somewhere? If not, it would be good to document them. Also mention what they are used for.